server/src/main/java/org/apache/druid/server/coordinator/SegmentReplicantLookup.java

```diff
@@ -114,6 +117,16 @@ public int getLoadedReplicants(SegmentId segmentId, String tier)
     return (retVal == null) ? 0 : retVal;
   }
 
+  public void setReplicationFactor(SegmentId segmentId, Integer requiredReplicas)
```

Comment:
It looks like `SegmentReplicantLookup` is following a singleton pattern with only getters, making it an immutable view by design (please correct me if I'm reading this wrong). Introducing a setter here and mutating the state when load rules are evaluated seems weird. I wonder if this information can be stored and fetched from a different place that's not immutable?

Reply:
That's true but `SegmentReplicantLookup` is the class which handles all replica-related info. Adding a new class just so that we can continue to keep `SegmentReplicantLookup` as immutable is probably not desirable.

Also, once the replication factor info is captured as a part of the `RunRules` coordinator duty, the `SegmentReplicantLookup` is not modified again. So we can potentially make an immutable copy at the end of the coordinator run.

But I prefer to keep the changes simple for this PR as this class is being fully revamped as a part of #13197 . In that PR, we can reconsider the immutability aspect of this class.

Reply:
@kfaraz, thanks for the clarification. That's fair, especially if it's revisited in the PR you linked. I'd suggest documenting this with a TODO so it's not confusing for someone reading this code.

Reply:
Added a comment for this.